extension/js/common/oauth2/oauth2.ts

```diff
@@ -23,10 +23,14 @@ export class OAuth2 {
     }
     const tabId = oauthWin?.tabs && oauthWin.tabs[0].id;
     return await new Promise((resolve) => {
-      BrowserMsg.addListener('auth_window_result', async (result: Bm.AuthWindowResult) => {
-        void chrome.tabs.remove(tabId!);
-        resolve(result);
-        return false;
+      // need to use chrome.runtime.onMessage because BrowserMsg.addListener doesn't work
+      // In gmail page reconnect auth popup, it sends event to background page (BrowserMsg.send.bg.await.reconnectAcctAuthPopup)
+      // thefore BrowserMsg.addListener doesn't work
+      chrome.runtime.onMessage.addListener((message: Bm.Raw) => {
+        if (message.name === 'auth_window_result') {
+          void chrome.tabs.remove(tabId!);
+          resolve(message.data.bm as Bm.AuthWindowResult);
+        }
```

Comment:
I'd like to understand the situation a little better, maybe there is another solution. Is it the following?

1) user needs to authenticate from Gmail content script, which doesn't have `chrome.windows` access
2) the content script calls background page
3) background page calls `newAuthPopup` which calls this `webAuthFlow` which produces an auth window
4) we have two listeners here in `webAuthFlow`
    - `chrome.tabs.onRemoved` which waits for the window to be closed by user
    - `auth_window_result` listener which waits for the `auth_window_result` message sent from the oauth window content script
5) user approves the auth, and the content script broadcasts `auth_window_result` message
6) the message is intercepted above which will resolve in this method
7) the backround page `reconnect_acct_auth_popup` call will resolve back into the gmail content script, which will now know that user was successfully authenticated

--------------------

On step 5, since it's a broadcast message, it's supposed to reach every frame. But you're saying it only reached the backround page and nowhere else? Sound like our `broadcast` mechanism is broken.

Reply:
Exactly!
chrome.runtime.onMessage works but BrowserMsg.addListener doesn't work here.
I also think `BrowserMsg.addListener` has some issues for this case.